src/tunnel.cc

```diff
@@ -898,6 +898,17 @@ void
 TunnelStateData::copyRead(Connection &from, IOCB *completion)
 {
     assert(from.len == 0);
+
+    // TODO: remove code duplication, creating a helper method
+    if (!Comm::IsConnOpen(from.conn)) {
```

Comment:
If from.conn is not open anymore, then why was not our connection closure callback called? And if it _was_ called, why do we need to do some extra steps here/now to end transaction and destroy TunnelStateData?

Reply:
Perhaps we can rely on our connection closure callback and simply return here. However, it looks like that current tunnel code is inconsistent in this respect. For example, TunnelStateData::writeServerDone() does not rely on the callback (under the "If the other end has closed, so should we" comment) but TunnelStateData::readClient() does (under the "close handlers will tidy up for us" comment). I could not find a place whether/where we remove close callbacks in tunnel.cc. So it looks like that "relying on close handlers" is the correct approach.